### PR TITLE
Deny price unit changes of in-use Shop (SHOOP-1658)

### PR DESCRIPTION
**Pull request by suutari-ai**
Changing currency or taxfulness of Shop's prices while it has any orders
is not good, so deny it.

Also rename ImmutableMixin to ChangeProtected and improve it a bit and
make the non-field errors visible in ShopEditView.

### CODE REVIEW

**Comment by suutari-ai** [comment]
Oh wait.. just noticed that this is not yet strictly to the spec... gotta update this..

**Comment by suutari-ai** [comment]
Actually the spec was stupid. Updated the issue description. This PR should be OK now.

File: shoop/core/models/_base.py [hunk 1]
```diff
@@ -47,28 +47,37 @@ class Meta:
         abstract = True
 
 
-class ImmutableMixin(object):
+class ChangeProtected(object):
+    protected_fields = None
     unprotected_fields = []
-    immutability_message = _("Cannot change immutable object that is in use")
+    change_protect_message = _("Following fields cannot be changed")
```
Comment (akx) [inline shoop/core/models/_base.py]: > The following fields can not be changed

**Comment by akx** [comment]
This is also a semver major change.

File: shoop/core/models/_base.py [hunk 2]
```diff
@@ -47,28 +47,37 @@ class Meta:
...
+                    fields=", ".join(sorted(changed_protected_fields)),
+                )
+                raise ValidationError(message)
 
     def save(self, *args, **kwargs):
         self.clean()
-        super(ImmutableMixin, self).save(*args, **kwargs)
+        super(ChangeProtected, self).save(*args, **kwargs)
 
     def _is_in_use(self):
```
Comment (akx) [inline shoop/core/models/_base.py]: This isn't really a private function? I mean, it's supposed to be overridden in subclasses..

**Comment by suutari-ai** [comment]
> This is also a semver major change.

How so?

File: shoop/core/models/_base.py (hunk 2 above)
Comment (suutari-ai) [inline shoop/core/models/_base.py]: I don't understand what you're after in this..

File: shoop/core/models/_base.py (hunk 1 above)
Comment (suutari-ai) [inline shoop/core/models/_base.py]: Updated.

File: shoop/core/models/_base.py (hunk 2 above)
Comment (akx) [inline shoop/core/models/_base.py]: It's an underscore-prefixed function, which hints that it's a private API, but subclasses of `ChangeProtected`s are supposed to override it. I think it should be just `is_in_use` (and it should have a proper docstring too).

**Comment by akx** [comment]
> > This is also a semver major change.
> How so?

The publicly usable name `ImmutableMixin` is gone.

**Comment by suutari-ai** [comment]
ImmutableMixin was never publicly usable. It's only available through an import from a private submodule (`shoop.core.models._base`).

**Comment by akx** [comment]
@suutari-ai Oh, yeah, you're right. I didn't pay attention to the file name there... Sorry!

File: shoop/core/models/_base.py (hunk 2 above)
Comment (suutari-ai) [inline shoop/core/models/_base.py]: This PR didn't touch it.

Anyway, something being underscored does not mean it's private. It means that it's non-public. Quoting from PEP8:

> Use one leading underscore only for non-public methods and instance variables.

Being non-public does not mean it's not supposed to be overridden, it just means it's not supposed to be called from outside the class hierarchy. I.e. underscoring hints more about being protected than private (in C++/Java terms). But, yeah this is not really well defined anywhere and people's expectation might vary. Good docstring would help much.

Comment (suutari-ai) [inline shoop/core/models/_base.py]: Renamed it and added a docstring. Better now?

**Pull request merged**